Nahom-Derese/Loan-Tracking-API
# Login and Signup Routes Implemented

## Description

This PR completes the implementation of the user login and signup routes for the Loan Tracker API. It includes the following:

### Signup Route

Allows new users to register by providing their details (first name, last name, email, password, phone number, and address).

### Login Route

Allows existing users to authenticate using their email and password, and returns a JWT token upon successful authentication.

### Validation

Input validation has been added for both routes to ensure data integrity.

### Error Handling

Proper error handling has been implemented to manage various failure scenarios such as incorrect passwords or existing email addresses.

#### Changes Implemented

#### User Registration Endpoint (POST /signup):

- Accepts user details through the RegisterUserForm.
- Validates the incoming data.
- Hashes the password using bcrypt.
- Stores the user details in the database.
- Returns a success message upon successful registration.

#### User Login Endpoint (POST /login):

- Accepts login credentials through the LoginUserForm.
- Validates the incoming data.
- Checks the provided password against the stored hashed password.
- Generates a JWT token upon successful authentication.
- Returns the JWT token in the response.

### Validation and Security:

- Added input validation to both signup and login forms using the go-playground/validator package.
- Passwords are hashed using bcrypt before being stored.
- Implemented JWT authentication for secure login.

### Middleware:

- Added a middleware to handle JWT authentication for protected routes (future implementation).

### Error Responses:

- Added clear and descriptive error messages for validation failures, authentication failures, and other common scenarios.

### How to Test

#### Run the Server:

- Start the server using `go run cmd/main.go`.

#### Test the Signup Route:

- Use Postman or a similar tool to send a POST request to `/signup` with a valid JSON body containing user details.
- Verify that the response indicates successful registration and that the user is stored in the database.

#### Test the Login Route:

- Use Postman or a similar tool to send a POST request to `/login` with valid login credentials.
- Verify that the response includes a JWT token.

#### Invalid Input Handling:

- Test the endpoints with invalid inputs (e.g., missing fields, invalid email format) and verify that appropriate error messages are returned.

### Next Steps

- Implement role-based access control (RBAC) using the JWT tokens.
- Protect sensitive routes using the JWT middleware.
- Add unit tests for the signup and login functionalities.

### Related Issues

#1 Set Up Project Structure and Dependencies
#2 Implement Error Handling and Logging Mechanisms
#3 Create Models for User and Loans
#4 Create Form Models for User Registration and Login
#5 Set Up MongoDB Connection and Configuration
#6 Implement User Login and Signup Endpoint with JWT Authentication
